runtime/nvqir/cutensornet/tensornet_state.cpp

```diff
@@ -0,0 +1,371 @@
+  // Destroy the quantum circuit sampler
+  HANDLE_CUTN_ERROR(cutensornetDestroySampler(sampler));
+  return counts;
+}
+
+std::vector<std::complex<double>> TensorNetState::getStateVector() {
+  LOG_API_TIME();
+  void *d_sv{nullptr};
+  const std::size_t svDim = 1u << m_numQubits;
+  HANDLE_CUDA_ERROR(cudaMalloc(&d_sv, svDim * sizeof(std::complex<double>)));
+  ScratchDeviceMem scratchPad;
+
+  // Create the quantum state amplitudes accessor
+  cutensornetStateAccessor_t accessor;
+  // Revert the bit-ordering
+  std::vector<int64_t> strides;
+  for (std::size_t bitIdx = 0; bitIdx < m_numQubits; ++bitIdx)
+    strides.emplace_back(1LL << (m_numQubits - bitIdx - 1));
+
+  HANDLE_CUTN_ERROR(cutensornetCreateAccessor(
+      m_cutnHandle, m_quantumState, 0, nullptr, strides.data(), &accessor));
+
+  const int32_t numHyperSamples =
+      8; // desired number of hyper samples used in the tensor network
+         // contraction path finder
+  HANDLE_CUTN_ERROR(cutensornetAccessorConfigure(
+      m_cutnHandle, accessor, CUTENSORNET_ACCESSOR_OPT_NUM_HYPER_SAMPLES,
+      &numHyperSamples, sizeof(numHyperSamples)));
+  // Prepare the quantum state amplitudes accessor
+  cutensornetWorkspaceDescriptor_t workDesc;
+  HANDLE_CUTN_ERROR(
+      cutensornetCreateWorkspaceDescriptor(m_cutnHandle, &workDesc));
+  HANDLE_CUTN_ERROR(cutensornetAccessorPrepare(
+      m_cutnHandle, accessor, scratchPad.scratchSize, workDesc, 0));
+
+  // Attach the workspace buffer
+  int64_t worksize = 0;
+  HANDLE_CUTN_ERROR(cutensornetWorkspaceGetMemorySize(
+      m_cutnHandle, workDesc, CUTENSORNET_WORKSIZE_PREF_RECOMMENDED,
+      CUTENSORNET_MEMSPACE_DEVICE, CUTENSORNET_WORKSPACE_SCRATCH, &worksize));
+  if (worksize <= static_cast<int64_t>(scratchPad.scratchSize)) {
+    HANDLE_CUTN_ERROR(cutensornetWorkspaceSetMemory(
+        m_cutnHandle, workDesc, CUTENSORNET_MEMSPACE_DEVICE,
+        CUTENSORNET_WORKSPACE_SCRATCH, scratchPad.d_scratch, worksize));
+  } else {
+    throw std::runtime_error("ERROR: Insufficient workspace size on Device!");
+  }
+
+  // Compute the quantum state amplitudes
+  std::complex<double> stateNorm{0.0, 0.0};
+  HANDLE_CUTN_ERROR(
+      cutensornetAccessorCompute(m_cutnHandle, accessor, nullptr, workDesc,
+                                 d_sv, static_cast<void *>(&stateNorm), 0));
+  std::vector<std::complex<double>> h_sv(svDim);
+  HANDLE_CUDA_ERROR(cudaMemcpy(h_sv.data(), d_sv,
+                               svDim * sizeof(std::complex<double>),
+                               cudaMemcpyDeviceToHost));
+
+  // Free resources
+  HANDLE_CUTN_ERROR(cutensornetDestroyWorkspaceDescriptor(workDesc));
+  HANDLE_CUTN_ERROR(cutensornetDestroyAccessor(accessor));
+  HANDLE_CUDA_ERROR(cudaFree(d_sv));
+
+  return h_sv;
+}
+
+std::vector<std::complex<double>>
+TensorNetState::computeRDM(const std::vector<int32_t> &qubits) {
+  LOG_API_TIME();
+  void *d_rdm{nullptr};
+  const std::size_t rdmSize = 1u << (2 * qubits.size());
```

Comment:
The bit shift operations will return unexpected results if we're shifting by too many bits, and we're using them in a handful of places. Are there any up-front error checks that ensure that we won't exceed any limits? I.e. in this case, having qubits.size() == 32 will cause unexpected behavior without throwing any immediate error messages.

Reply:
I've added a check in those places so that we don't accidentally overflow the number and cause unexpected failures.